cpp/oneapi/dal/detail/array_impl.hpp

```diff
@@ -163,22 +162,47 @@ class array_impl : public base {
     template <typename Y>
     void reset(const array_impl<Y>& ref, T* data, std::int64_t count) {
         if (ref.has_mutable_data()) {
-            data_owned_ = shared(std::get<1>(ref.data_owned_), data);
+            if (const auto& ptr = std::get_if<1>(&ref.data_owned_)) {
+                data_owned_ = shared(*ptr, data);
+            }
+            else {
+                throw internal_error(
+                    dal::detail::error_messages::array_does_not_contain_mutable_data());
+            }
         }
         else {
-            data_owned_ = shared(std::get<0>(ref.data_owned_), data);
+            if (const auto& ptr = std::get_if<0>(&ref.data_owned_)) {
+                data_owned_ = shared(*ptr, data);
+            }
+            else {
+                throw internal_error(
```

Comment:
I'd suggest changing it to a more meaningful error message that the reference array is empty and does not contain ownership structure.

Reply:
applied